### ISSUE TITLE
mega: account lockouts after protocol changes on server side

### ISSUE THREAD

**Issue by syedrali** [rclone/rclone#3393]
Using rclone to sync with MEGA has started causing my MEGA account to be locked due to a recent change on the MEGA side. This is what MEGA support said:

"The rclone software do not fully implement our cryptographic protocols, they sent the login request with a old credential logic, that is not secure. So to prevent any damage your account was blocked.
They have to upgrade they credential logic and implement our cryptographic protocols to avoid that security measure."

**Comment by ncw** [comment]
Which version of rclone are you using?  We changed the way the mega login works in v1.46

**Comment by syedrali** [comment]
I don't know. I followed the instructions on https://rclone.org/install/ which points at https://rclone.org/install.sh . Is this not the correct version?

**Comment by syedrali** [comment]
rclone version
rclone v1.48.0
- os/arch: linux/amd64
- go version: go1.12.6

**Comment by syedrali** [comment]
This started happening today after regular automated rclone usage, and since my rclone version appears up to date, I think it is due to a change on the Mega end?

**Comment by ncw** [comment]
That is a pain :-(

The mega API isn't documented and the only way of figuring it out is reading their open source C++ program which is pretty hard work!

Can you ask Mega what was changed and when?

**Comment by syedrali** [comment]
They've closed out the support ticket now I'm afraid. That was all they would say and that they didn't support third-party tools telling me not to use rclone.

**Comment by ncw** [comment]
Did rclone start working again?  It works when I try it here.

**Comment by syedrali** [comment]
Not sure. I haven't tried it as Mega Support were very rude about my rclone usage saying it was why my account had been locked as quoted above. So I've been decoupling myself from Mega.

**Comment by ncw** [comment]
Fair enough!  I'll leave this issue open to see if anyone else is having the same problem.

**Comment by msalini** [comment]
I have to report similar problems with Mega too, however it is intermittent.
I have 3 different accounts with Mega, and only one gives me problems. No matter how many times I delete and re-create the entry in the config file, It won't work. Here are the details for the problematic one::

``  rclone -vv about Mega:
2019/10/22 13:22:37 DEBUG : rclone: Version "v1.49.5" starting with parameters ["rclone" "-vv" "about" "Mega:"]
Enter configuration password:
password:
2019/10/22 13:22:44 DEBUG : Using config file from "/home/xxxxxxxxxxx/.config/rclone/rclone.conf"
2019/10/22 13:22:57 Failed to create file system for "Mega:": couldn't login: Try again
``

I am using Linux Mint based on Ubuntu 16.04.

**Comment by ncw** [comment]
I've heard reports of other mega accounts which get locked out for days then start working again.  Hopefully that will happen to yours @msalini

**Comment by msalini** [comment]
> I've heard reports of other mega accounts which get locked out for days then start working again. Hopefully that will happen to yours @msalini

@ncw I am using a temporary hack at the moment: I have installed MegaCMD from the mega.nz website. I have created a webdav instance with MegaCMD on localhost and I use rclone to connect to it via webdav. Not exactly the most elegant solution but at least I do not need to change all my current scripts.

**Comment by ncw** [comment]
Great workaround!  How well does the webdav instance work?

**Comment by msalini** [comment]
> Great workaround! How well does the webdav instance work?

It is super-fast!
I almost feels like browsing a local drive :-)
I was actually considering whether you could potentially embed this solution in rclone itself. Or something similar. Given that MegaCMD is an official (and open source) product by Mega, this would resolve all the ongoing issues with them.

**Comment by ncw** [comment]
> > Great workaround! How well does the webdav instance work?
> 
> It is super-fast!
> It almost feels like browsing a local drive :-)
> I was actually considering whether you could potentially embed this solution in rclone itself. Or something similar. Given that MegaCMD is an official (and open source) product by Mega, this would resolve all the ongoing issues with them.

It is tempting, I have to admit.  The mega backend has been a world of pain...

How do you set up the webdav server - can you post a quick guide?

Maybe it is something rclone could do "on the fly" for a a mega/webdav interface.

**Comment by msalini** [comment]
I used the interactive interface of MegaCMD, but thisd can also be scripted within MegaCMD.
In short:
1. I opened the interactive session of MegaCMD: `` mega-cmd``
2. I logged in with my credentials: `` MEGA CMD> signup [user] [password] ``
3. I checked that my root in the Mega filesystem is "/" using the `` mount`` command
4. I initiated a webdav session on localhost: 
`` $ webdav / --port=9090                              
Serving via webdav /: http://127.0.0.1:9090/wBwBEDjZ/Cloud%20Drive ``

Then I set up rclone to connect to localhost via webdav (no username, no password, "other" webdav):

>  e) Edit existing remote
n) New remote
d) Delete remote
r) Rename remote
c) Copy remote
s) Set configuration password
q) Quit config
e/n/d/r/c/s/q> n
name> MegaWebDav
Type of storage to configure.
Enter a string value. Press Enter for the default ("").
Choose a number from below, or type in your own value
 ...... 
Storage> 28
** See help for webdav backend at: https://rclone.org/webdav/ **
URL of http host to connect to
Enter a string value. Press Enter for the default ("").
Choose a number from below, or type in your own value
 1 / Connect to example.com
   \ "https://example.com"
url> http://127.0.0.1:9090/wBwBEDjZ/Cloud%20Drive
Name of the Webdav site/service/software you are using
Enter a string value. Press Enter for the default ("").
Choose a number from below, or type in your own value
 1 / Nextcloud
   \ "nextcloud"
 2 / Owncloud
   \ "owncloud"
 3 / Sharepoint
   \ "sharepoint"
 4 / Other site/service or software
   \ "other"
vendor> 4
User name
Enter a string value. Press Enter for the default ("").
user> 
Password.
y) Yes type in my own password
g) Generate random password
n) No leave this optional password blank
y/g/n> n
Bearer token instead of user/pass (eg a Macaroon)
Enter a string value. Press Enter for the default ("").
bearer_token> 
Edit advanced config? (y/n)
y) Yes
n) No
y/n> n
Remote config
--------------------
>[MegaWebDav]
type = webdav
url = http://127.0.0.1:9090/wBwBEDjZ/Cloud%20Drive
vendor = other
--------------------

Hope this is clear enough.
I think you can programatically pass the Mega username and password to mega-cmd and initiate the webdav session before setting up the rclone instance. you are trading a dependency on mega-cmd for the advantage of not having to deal with the unpredictability of the Mega "secret" APIs.

**Comment by ncw** [comment]
@msalini thanks for that.  I had a go and I have to say it does work very well.  The mega protocol is designed to work best with a persistent daemon which has always been running slightly against what rclone does.

What I could do is make a "mega" mode for the webdav backend which can invoke mega-cmd automatically.

I don't think I'd remove the mega backend, maybe just deprecate it.

More experiments needed!

This should probably go in a new issue...

**Comment by msalini** [comment]
I thought along the same line. Transitionally, when a new Mega instance is created in rclone:

1. rclone checks whether mega-cmd is installed
2. If so, it starts up the mega-cmd demon and creates a webdav instance

If mega-cmd is not installed, it can initiate a traditional mega instance with a warning.

If I had some programming knowledge, I would gladly chip in, but I am just a humble user. Yet, glad to be able to contribute to your project :-)

**Comment by ncw** [comment]
@msalini great idea :-)

> If I had some programming knowledge, I would gladly chip in, but I am just a humble user. Yet, glad to be able to contribute to your project :-)

Actually if you'd like to summarize the idea of adding a mega mode to the webdav backend into a new issue that would be a great contribution - thank you :-)

**Comment by msalini** [comment]
[Done!](https://github.com/rclone/rclone/issues/3660) :-)

**Comment by caspertone2003** [comment]
I had in the past such an issue with a remote. After contactin with mega support, I got a nice ticket. No reply ever but later it started to work as normal.
I closed it for other reasons.
Now, in a new opened remote, I have once again same problem... I can just list root contents if I wait for some days, then it gets blocked for some other days.
Vote for an alternate mega-cmd based alternative to connect to mega....
Thanks!

**Comment by caspertone2003** [comment]
MegaCMD has a learning curve. Just a simple How To... for windows users ... batch mode

Once MegaCMD is installed etc...

MEGAclient.exe login useremailofremote password
MEGAclient.exe webdav /

it produces a URL such as http://127.0.0.1:4443/yourwhatever/Cloud%20Drive

Open explorer, map network drive, our your URL select letter and ready

**Comment by ivandeex** [comment]
@syedrali @msalini 
This bug is reported against obsolete version of rclone. Can you reproduce it with rclone 1.54?

**Comment by ivandeex** [comment]
All in all this looks more like a provider issue.

@syedrali @caspertone2003 

Most probably your support requests are gone, but maybe you have links to the tickets?
